L1 cdf creation

bryan-harter opened the pull request:
# Change Summary

## Overview
Adding the ability to create CDF files from the L1 xarray data for IDEX

## New Dependencies
cdflib 

## New Files
- global_cdf_attrs.py
   - A place to store all common CDF attribute dictionaries
- idex_cdf_attrs.py
  - A place to store all IDEX specific CDF attribute dictionaries 

## Updated Files
- idex_packet_parser.py
   - Added the function "write_l1_cdf".  When called it generates a l1 CDF file from the L1 data

## Testing
- test_l1_cdfs.py
  - Unit tests for the CDF file creation

bryan-harter committed:
Extracting and adding new data to the IDEX L1

bryan-harter committed:
Adding some of the values to the XML, rather than shifting bits in the code
(work in progress, need to do a few more variables)

bryan-harter committed:
Finishing up having space packet parser decom

bryan-harter committed:
Getting rid of unecessary bit masksi8kt9gloh

bryan-harter committed:
making changes based on comments

bryan-harter committed:
Merge branch 'dev' into idex_l1_cdfs

bryan-harter committed:
Formatting the file after the upstream sync

bryan-harter committed:
Merge branch 'IMAP-Science-Operations-Center:dev' into idex_l1_cdfs

bryan-harter committed:
Adding L1 CDF creation

bryan-harter committed:
Forgot to commit poetry changes

bourque commented:
@all-contributors please add @bryan-harter for Code!

greglucas commented on imap_processing/common_cdf_attrs.py:
```diff
@@ -0,0 +1,29 @@
+import numpy as np
+
+global_base = {
+    "Project": ["STSP Cluster>Solar Terrestrial Science Programmes, Cluster"],
```
Why do we have Cluster here?

I would have thought it'd be "STP>Solar Terrestrial Probes"
https://science.nasa.gov/heliophysics/programs/solar-terrestrial-probes

bryan-harter replied:
Fixed

greglucas commented on imap_processing/common_cdf_attrs.py:
```diff
@@ -0,0 +1,29 @@
+    "Project": ["STSP Cluster>Solar Terrestrial Science Programmes, Cluster"],
+    "Source_name": ["IMAP>Interstellar Mapping and Acceleration Probe"],
+    "Discipline": ["Solar Physics>Heliospheric Physics"],
+    "Descriptor": ["IDEX>Interstellar Dust Experiment"],
```
This should only be in the idex group I think?

bryan-harter replied:
Fixed

greglucas commented on imap_processing/common_cdf_attrs.py:
```diff
@@ -0,0 +1,29 @@
+    "Source_name": ["IMAP>Interstellar Mapping and Acceleration Probe"],
+    "Discipline": ["Solar Physics>Heliospheric Physics"],
+    "Descriptor": ["IDEX>Interstellar Dust Experiment"],
+    "PI_name": ["Dave McComas"],
```
```suggestion
    "PI_name": ["David J. McComas"],
```
I assume we should use full qualified names?

bryan-harter replied:
Fixed, and also added "Dr." because thats how the one IBEX CDF file I found had it

greglucas commented on imap_processing/common_cdf_attrs.py:
```diff
@@ -0,0 +1,29 @@
+    "Discipline": ["Solar Physics>Heliospheric Physics"],
+    "Descriptor": ["IDEX>Interstellar Dust Experiment"],
+    "PI_name": ["Dave McComas"],
+    "PI_affiliation": ["Princeton"],
```
Princeton University?

bryan-harter replied:
Yeah I added more detailed here, including an address

greglucas commented on imap_processing/common_cdf_attrs.py:
```diff
@@ -0,0 +1,29 @@
+    "PI_name": ["Dave McComas"],
+    "PI_affiliation": ["Princeton"],
+    "Instrument_type": ["Particles (space)"],
+    "Mission_group": ["IMAP"],
```
We put IMAP here, but we never define IMAP, do you need to say what IMAP stands for somewhere?

greglucas commented on imap_processing/common_cdf_attrs.py:
```diff
@@ -0,0 +1,29 @@
+epoch_attrs = {
+    "CATDESC": "Default time",
+    "FIELDNAM": "Epoch",
+    "FILLVAL": np.array([-9223372036854775808]),
```
```suggestion
    "FILLVAL": np.iinfo(np.int64).min,
```
Why are these an array, should they be just a single value? Also might help to say it is the minimum possible int64?

Same for the other values below this.

greglucas commented on imap_processing/idex/idex_cdf_attrs.py:
```diff
@@ -0,0 +1,148 @@
+import numpy as np
+
+from .. import common_cdf_attrs
```
```suggestion
from imap_processing import common_cdf_attrs
```

greglucas commented on imap_processing/idex/idex_cdf_attrs.py:
```diff
@@ -0,0 +1,148 @@
+from .. import common_cdf_attrs
+
+# Set IDEX software version here for now
+software_version = "01"
```
```suggestion
__version__ = "01"
```
Do we want to follow more standard Python version naming schemes here?

greglucas commented on imap_processing/idex/idex_cdf_attrs.py:
```diff
@@ -0,0 +1,148 @@
+        "Data_type": ["l1"],
+        "Data_version": [software_version],
+        "TEXT": [
+            """Greg Newcomb, the flight software (FSW) engineer for the Interstellar
```
This is going into every file? This seems more like a test/verification note.

greglucas commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -3,10 +3,14 @@
 from imap_processing import imap_module_directory
 
+from .. import common_cdf_attrs
+from . import idex_cdf_attrs
```
```suggestion
from imap_processing import common_cdf_attrs
import idex_cdf_attrs
```

greglucas commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -89,6 +93,48 @@ def __init__(self, packet_file: str):
 
         self.data = xr.concat(processed_dust_impact_list, dim="Epoch")
 
+    def write_l1_cdf(self, description: str = "", version: str = "01"):
```
Is this version different from the software version? Would help to clarify what this version is based upon.

bryan-harter replied:
It isn't really based on anything at the moment, it just writes the version number to the end of the file name.  It should probably be based on the IDEX software version, but I didn't think we had reached any conclusions yet about how software versions corresponded to file versions, so I left it open-ended for now.

greglucas commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -89,6 +93,48 @@ def __init__(self, packet_file: str):
+
+        # Convert the xarray object to a CDF!
+
+        try:
+            xarray_to_cdf(
+                self.data, filename, terminate_on_warning=True
+            )  # Terminate if not ISTP compliant
+        except Exception as e:
+            logging.error("Error writing a CDF file! Aborting....")
+            logging.error(str(e))
+            return
```
Why not just let this raise? I don't think we need this in a try/except. You say terminate, but then you just return from the exception instead of re-raising.

bryan-harter replied:
I'll change that. I initially just figured if we wanted to write multiple IDEX files we didn't want the whole program to crash.  But we can just have the try/except located outside of whatever calls this function in that case.

greglucas commented on imap_processing/idex/tests/test_l1_cdfs.py:
```diff
@@ -0,0 +1,79 @@
+
+
+def test_idex_cdf_file(decom_test_data):
+    # Verify that a CDF file can be created with no errors thrown by xarray_to_cdf
+    file_name = decom_test_data.write_l1_cdf(version="01", description="")
+    assert file_name == "imap_idx_l1_20250724_0000_p_v01.cdf"
+    os.remove(file_name)
```
If a failure were to happen this would leave files around in the current directory. Should your writer take a directory path as well for where to put the file?
https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory

```python
with tempfile.TemporaryDirectory() as tmpdir:
    ...
```
which will clean up for you properly.

This also applies to all the other tests as well.

bryan-harter replied:
Added

greglucas commented on imap_processing/idex/tests/test_l1_cdfs.py:
```diff
@@ -0,0 +1,79 @@
+
+def test_bad_cdf_attributes(decom_test_data):
+    # Deliberately mess up the attributes to verify that no CDF is created
+    copy_to_modify = copy.deepcopy(decom_test_data)
```
Rather than needing to deepcopy this, we should just update the fixture above to not be `session`, so it will create a new data file every single time if you need a fresh one.

bryan-harter replied:
changed

greglucas reviewed (commented):
Overall looks like a good way to handle the writing to me.

bryan-harter reviewed (commented):
I apparently need to submit a review for my comments to go through

bryan-harter committed:
Making changes based on Greg's comments,
as well as issues 1-8 from SPDF

bryan-harter committed:
Fixing import statements

bryan-harter committed:
Changing version into a string

maxinelasp commented on imap_processing/common_cdf_attrs.py:
```diff
@@ -0,0 +1,31 @@
+    "PI_name": ["Dr. David J. McComas"],
+    "PI_affiliation": [
+        "Princeton Plasma Physics Laboratory",
+        "100 Stellarator Road, Princeton, NJ 08540",
```
I wish I lived on Stellarator road...

maxinelasp commented on imap_processing/idex/idex_cdf_attrs.py:
```diff
@@ -0,0 +1,149 @@
+    {
+        "Data_type": ["l2"],
+        "Data_version": [__version__],
+        "TEXT": [
```
Could we move the common attributes into one variable? Or have some kind of `base_dict` which you can then expand?

maxinelasp commented on imap_processing/idex/idex_cdf_attrs.py:
```diff
@@ -0,0 +1,149 @@
+    "FORMAT": "E12.2",
+    "UNITS": "dN",
+    "VALIDMIN": np.array([-1.0e31], dtype=np.float32),
+    "VALIDMAX": np.array([1.0e31], dtype=np.float32),
```
Similarly, can all these magic numbers get moved to a constant somewhere and maybe commented on why those values were chosen?

maxinelasp commented on imap_processing/idex/idex_cdf_attrs.py:
```diff
@@ -0,0 +1,149 @@
+    | common_cdf_attrs.global_base
+)
+
+# L1 variables base dictionaries (these are not complete)
```
What does it take for these to be complete? Is it complete enough for others to use as an example (because it's only missing IDEX specific stuff) or are there other reasons?

I don't mind if you merge them incomplete btw, just curious what is left.

bryan-harter replied:
They're not "complete" in that they can't used as variable attributes as it is, they need more info and get completed by the dictionaries below it

maxinelasp commented on imap_processing/idex/idex_cdf_attrs.py:
```diff
@@ -0,0 +1,149 @@
+    "CATDESC": "TOF_Mid",
+    "FIELDNAM": "TOF_Mid",
+    "VAR_NOTES": "This is the mid gain channel of the time-of-flight signal.",
+} | l1_tof_base
```
I didn't know you could use `|` like this. Good new info!

tech3371 replied:
I was thinking that as I was reading above code!

bryan-harter committed:
Adding more descriptions, units, and labels in the attributes

bryan-harter committed:
Merge branch 'dev' into l1_cdf_creation

bryan-harter committed:
Fixing the way strings are inserted into the CDF

bryan-harter committed:
Fixing issues with white spice in the text of attrs

maxinelasp reviewed (commented):
I apparently didn't submit this two days ago. If the comments are outdated, just ignore them

tech3371 commented on imap_processing/idex/idex_cdf_attrs.py:
```diff
@@ -0,0 +1,167 @@
+from imap_processing import common_cdf_attrs
+
+# Set IDEX software version here for now
+__version__ = "1"
```
```suggestion
from imap_processing import common_cdf_attrs
from imap_processing.idex import version
```
Then in `imap_processing/idex/__init__.py`
```
version = "01"
```
I feel like we will use this variable a lot in different places. This way, it makes it easy to import anywhere. What do you think?

tech3371 commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -89,6 +91,45 @@ def __init__(self, packet_file: str):
+        file_start_date = self.data["Epoch"][0].data
+        date_string = np.datetime_as_string(file_start_date, unit="D").replace("-", "")
+        if description:  # Just append an underscore to the description
+            description = f"_{description}"
```
To avoid double underscore
```suggestion
            description = description if description.startswith("_") else f"_{description}"
```

tech3371 commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -89,6 +91,45 @@ def __init__(self, packet_file: str):
+        filename_and_path = os.path.join(directory, filename)
+        # Set the global attributes in the xarray
+        self.data.attrs = idex_cdf_attrs.idex_l1_global_attrs
+        # The Logical_file_id is always the name of the file without the extension
```
good to know

tech3371 commented on imap_processing/idex/tests/test_l1_cdfs.py:
```diff
@@ -0,0 +1,91 @@
+        file_name = decom_test_data.write_l1_cdf(
+            version="01", description="impact-lab-test001", directory=tmpdir
+        )
+        assert file_name == os.path.join(
+            tmpdir, "imap_idx_l1_20250724_0000_impact-lab-test001_p_v01.cdf"
+        )
```
How did this check that there is no CDF file? out of curiosity

greglucas replied:
Agreed, this should be something like: `assert not file_name.exists()`, but also do we want it to just continue on or should we raise an error/emit a warning if nothing gets created?

tech3371 reviewed (commented):
Looks really good and complicated! 

Can you write a guidance of how to write instrument specific attrs dictionary. Yours had lot of good information and I wouldn't know where to get those informations. Does those instrument specific attrs stores information about xarray.DataArray's dimension?

greglucas commented on imap_processing/common_cdf_attrs.py:
```diff
@@ -0,0 +1,31 @@
+import numpy as np
+
+global_base = {
+    "Project": ["STP>Solar-Terrestrial Physics"],
```
I find it odd that these are all lists. Is that what it is supposed to be or are they supposed to be single entities?

`Project` is singular, so should it be `Projects` as the lookup to indicate that these need to be lists? Will `PI_affiliation` below be two entries, or a single concatenated string? That reads as two entries currently, but I think they are related with name/address.

bryan-harter replied:
From cdflib's perspective it doesn't matter if they are lists or not, cdflib will insert them into a list if they aren't already. This just saves cdflib a step. 

Each member of the list is an "entry" of the attribute. Attributes can have multiple entries. The SPDF asked me to put the address as a separate entry of "PI_Affiliation".  

"Project" is the ISTP standard attribute, and their tools will complain if given the attribute name "Projects", even if there are multiple projects the data comes from.

greglucas commented on imap_processing/common_cdf_attrs.py:
```diff
@@ -0,0 +1,31 @@
+epoch_attrs = {
+    "CATDESC": "Default time",
+    "FIELDNAM": "Epoch",
+    "FILLVAL": np.iinfo(np.int64).min,
```
I feel like a lot of my comments are getting lost when I say to update them here and also below, you got here, but not the subsequent ones where this same pattern would be helpful (removing magic numbers as @maxinelasp pointed out as well). Would it help to comment in every place? I find that to be a lot of noise, but I'm happy to do it too if it'd be helpful.

greglucas commented on imap_processing/idex/tests/test_l1_cdfs.py:
```diff
@@ -0,0 +1,91 @@
+            version="01", description="", directory=tmpdir
+        )
+        assert file_name == os.path.join(tmpdir, "imap_idx_l1_20250724_0000_p_v01.cdf")
+        os.remove(file_name)
```
you don't need to remove anything, that is the nice thing about the context manager here.
```suggestion
```

greglucas commented on imap_processing/idex/tests/test_l1_cdfs.py:
```diff
@@ -0,0 +1,91 @@
+        file_name = decom_test_data.write_l1_cdf(
+            version="01", description="", directory=tmpdir
+        )
+        assert file_name == os.path.join(tmpdir, "imap_idx_l1_20250724_0000_p_v01.cdf")
```
This is a bit cryptic how you know that this is the proper filename. Should we add some of these identifiers to the `write_l1_cdf` routine so that routine can take in dates, software versions? Or another possibility is to make the string here be composed from those attributes `imap_idx_{level}_{date}_{version?}_{identifer?}_{version?}`
```suggestion
        assert file_name == str(tmp_path / "imap_idx_l1_20250724_0000_p_v01.cdf")
        assert Path(file_name).exists()
```
We should also assert that the file exists?

greglucas commented on imap_processing/idex/tests/test_l1_cdfs.py:
```diff
@@ -0,0 +1,91 @@
+
+def test_idex_cdf_file(decom_test_data):
+    # Verify that a CDF file can be created with no errors thrown by xarray_to_cdf
+    with tempfile.TemporaryDirectory() as tmpdir:
```
I just realized you can actually use the `tmpdir` or `tmp_path` fixture so we don't even need to do this.
https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html

```suggestion
def test_idex_cdf_file(decom_test_data, tmp_path):
    # Verify that a CDF file can be created with no errors thrown by xarray_to_cdf
```

greglucas commented on imap_processing/idex/tests/test_l1_cdfs.py:
```diff
@@ -0,0 +1,91 @@
+    # Verify that a sample of the data is correct inside the CDF file
+    # impact_14_tof_high_data.txt has been verified correct by the IDEX team
+    with open("imap_processing/idex/tests/impact_14_tof_high_data.txt") as f:
+        data = np.array([int(line.rstrip("\n")) for line in f])
```
rstrip() should remove the trailing newline by default?
```suggestion
        data = np.array([int(line.rstrip()) for line in f])
```

greglucas commented on imap_processing/common_cdf_attrs.py:
I don't love this name. What about just calling it `imap_processing/cdf_utils.py` or something like that? Do we want an entire `cdf` module with additional utilities? `cdf/global_attributes.py`, `cdf/utils.py`, all nested under a single directory.

greglucas reviewed (commented):
I don't quite get why the other comments are getting buried in the main thread and not showing up in the code review area for me... I see there are a lot of other comments, but I only see them under the conversation tab. I'll try and look into this...

bryan-harter committed:
Files are finally ISTP compliant!

bryan-harter committed:
Fixing a couple things for ISTP

bryan-harter committed:
Had the contents of the 32 bit things reversed

bryan-harter committed:
Fixing the packet definition for the CCSDS file

bryan-harter committed:
More fixes to ensure ISTP compliance

bryan-harter committed:
Updating the file name

bryan-harter committed:
touching finishes on the L1 CDF

bryan-harter committed:
Adding fixture for temp directory in tests

bryan-harter committed:
Fixing one of the ruff errors by adding a namedtuple

bryan-harter committed:
Adding new "write_cdf" function that writed cdfs based on attributes

bryan-harter committed:
Getting rid of a few more magic numbers

bryan-harter commented:
Ok I think I finally responded to all comments, and the code now makes files that are PDS compliant.  Summary of the major changes since a few weeks ago:

- Renamed "cdf_common_attrs.py" to "cdf_utils.py"
- Added a function "write_cdf()" to the cdf_utils, which writes a CDF file based on the global attributes of the xarray object
- Cleaned up the unit tests 
- Commented or replaced some of the magic numbers
- Modified quite a few of the values in the attribute dictionaries based on feedback from the SPDF

I also made a couple tweaks to cdflib.xarray_to_cdf, so the required cdflib version is now 1.2.3

bryan-harter committed:
Merge branch 'dev' into l1_cdf_creation

tech3371 commented on imap_processing/cdf_utils.py:
```diff
@@ -0,0 +1,102 @@
+}
+
+
+def write_cdf(data: xr.Dataset, description: str = "", directory: str = ""):
```
I am writing similar function in my working branch to make it more generic that we all can use. Later, I can combine this and mine into one and test it for boths.

tech3371 commented on imap_processing/cdf_utils.py:
```diff
@@ -0,0 +1,102 @@
+    data.attrs["Logical_file_id"] = filename.split(".")[0]
+
+    # Convert the xarray object to a CDF
+    xarray_to_cdf(
```
Did you notice that it doesn't overwrite existing file with new changes? I noticed that when I was testing it locally. I had to delete old files and run again to see new changes.

bryan-harter committed:
Updating the packet definition
